### PR TITLE
Fragment cache

### PR DESCRIPTION
**Pull request by jbellis**

### CODE REVIEW

**Comment by jbellis** [comment]
I'm not worrying about the case where the graph is too small to benefit from caching, b/c C* already tries to detect tiny graphs and do a brute force comparison anyway.  Can revisit later if necessary.

File: jvector-base/src/main/java/io/github/jbellis/jvector/pq/CompressedDecoder.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+                    int offset = pq.subvectorSizesAndOffsets[i][1];
+                    float[] centroidSubvector = pq.codebooks[i][j];
+                    switch (vsf) {
+                        case DOT_PRODUCT:
+                            cache[i][j] = VectorUtil.dotProduct(centroidSubvector, 0, centeredQuery, offset, centroidSubvector.length);
+                            break;
+                        case EUCLIDEAN:
+                            cache[i][j] = VectorUtil.squareDistance(centroidSubvector, 0, centeredQuery, offset, centroidSubvector.length);
+                            break;
+                        default:
```
Comment (jkni) [inline jvector-base/src/main/java/io/github/jbellis/jvector/pq/CompressedDecoder.java:38]: cosine too expensive to be worth precomputing?

File: jvector-base/src/main/java/io/github/jbellis/jvector/pq/CompressedDecoder.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+                            throw new UnsupportedOperationException("Unsupported similarity function " + vsf);
+                    }
+                }
+            }
+            return cache;
+        }
+
+        @Override
+        public float similarityTo(int node2) {
+            return (1 + decodedSimilarity(cv.get(node2), cache)) / 2;
```
Comment (jkni) [inline jvector-base/src/main/java/io/github/jbellis/jvector/pq/CompressedDecoder.java]: This changes the Euclidean similarity definition. Is that intentional?

Comment (jbellis) [reply jvector-base/src/main/java/io/github/jbellis/jvector/pq/CompressedDecoder.java]: No, thanks for the catch.

**Pull request merged**